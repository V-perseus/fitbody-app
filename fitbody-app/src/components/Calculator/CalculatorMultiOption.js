import React, { memo } from 'react'
import { StyleSheet, Text, TouchableOpacity, View, Keyboard } from 'react-native'

// Assets
import globals from '../../config/globals'

const CalculatorMultiOption = (props) => {
  /**
   * Set the active state
   */
  const handleActiveOption = (activeOption) => () => {
    Keyboard.dismiss()
    props.update(activeOption)
  }

  /**
   * Render out each option button
   */
  const renderOption = (data, index) => {
    const elements = props.elementsPerRow ?? 2
    const realWidth = (globals.window.width * 0.86) / elements
    const firstElementStyle = props.type === 'pregnancy' && index === 0 ? { width: '100%' } : undefined
    const active = data.value === props.activeOption

    return (
      <TouchableOpacity
        key={`option-${index}`}
        style={[styles.button, active ? styles.activeButton : styles.inactiveButton, { width: realWidth }, firstElementStyle]}
        onPress={!active ? handleActiveOption(data.value) : null}>
        <View style={styles.buttonContainer}>
          <Text style={[styles.titleLabel, active ? styles.activeLabel : styles.inactiveLabel]}>{data.title.toUpperCase()}</Text>
          {data.subtitle ? <Text style={active ? styles.activeSubtitleLabel : styles.subtitleLabel}>{data.subtitle}</Text> : null}
        </View>
      </TouchableOpacity>
    )
  }

  return (
    <View style={styles.container}>
      <Text style={styles.label}>{props.title}</Text>
      {props.type === 'pregnancy' ? (
        <View style={styles.rowPregnancy}>{props.options.map(renderOption)}</View>
      ) : (
        <View style={styles.row}>{props.options.map(renderOption)}</View>
      )}
    </View>
  )
}

export default memo(CalculatorMultiOption)

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'center',
    width: globals.window.width,
    paddingLeft: 20,
    paddingRight: 20,
    marginTop: 20,
    marginBottom: 20,
  },
  label: {
    color: globals.styles.colors.colorBlack,
    fontSize: 16,
    fontFamily: globals.fonts.primary.semiboldStyle.fontFamily,
  },
  row: {
    flex: 2,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    flexWrap: 'wrap',
    flexGrow: 0,
  },
  rowPregnancy: {
    flex: 2,
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    flexWrap: 'wrap',
    flexGrow: 0,
  },
  button: {
    height: 50,
    width: globals.window.width * 0.43,
    borderRadius: 30,
    marginBottom: 10,
    marginTop: 10,
  },
  buttonContainer: {
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
    height: 50,
  },
  activeButton: {
    backgroundColor: globals.styles.colors.colorPink,
  },
  inactiveButton: {
    borderColor: globals.styles.colors.colorGray,
    borderWidth: 1,
  },
  activeLabel: {
    fontSize: 12,
    color: globals.styles.colors.colorWhite,
  },
  inactiveLabel: {
    fontSize: 12,
    color: globals.styles.colors.colorBlack,
  },
  titleLabel: {
    color: globals.styles.colors.colorBlack,
    fontFamily: globals.fonts.primary.semiboldStyle.fontFamily,
    textAlign: 'center',
  },
  subtitleLabel: {
    fontSize: 10,
    color: globals.styles.colors.colorGrayDark,
  },
  activeSubtitleLabel: {
    fontSize: 10,
    color: globals.styles.colors.colorWhite,
  },
})
