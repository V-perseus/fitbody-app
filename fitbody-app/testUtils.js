import React from 'react'
import { render as rtlRender } from '@testing-library/react-native'
import { configureStore } from '@reduxjs/toolkit'
import { NavigationContainer } from '@react-navigation/native'
import { createStackNavigator } from '@react-navigation/stack'
import { Provider } from 'react-redux'

// Import your own reducer
import { appReducer } from './src/store'
import { ModalStackScreens } from './src/config/routes/loggedInStack'

// custom render method override
function render(ui, { preloadedState, store = configureStore({ reducer: appReducer, preloadedState }), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react-native'
// override render method
export { render }

// custom navigation mock
const Stack = createStackNavigator()
export const MockedNavigator = ({
  component,
  params = {},
  preloadedState,
  store = configureStore({ reducer: appReducer, preloadedState }),
}) => {
  return (
    <Provider store={store}>
      <NavigationContainer>
        <Stack.Navigator>
          <Stack.Screen name="MockedScreen" component={component} initialParams={params} />
          <Stack.Screen name="Modals" component={ModalStackScreens} />
        </Stack.Navigator>
      </NavigationContainer>
    </Provider>
  )
}
