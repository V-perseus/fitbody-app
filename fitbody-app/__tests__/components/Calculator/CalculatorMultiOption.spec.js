import React from 'react'
import { fireEvent, render, cleanup } from '@testing-library/react-native'
import { render as appRender } from '../../../testUtils'

import CalculatorMultiOption from '../../../src/components/Calculator/CalculatorMultiOption'

afterEach(cleanup)

describe('CalculatorMultiOption', () => {
  const onPress = jest.fn()
  const options = [
    {
      value: 1,
      title: 'title_1',
      subtitle: 'subtitle_1',
    },
    {
      value: 2,
      title: 'title_2',
      subtitle: 'subtitle_2',
    },
  ]

  it('renders snapshot correctly', () => {
    const tree = appRender(<CalculatorMultiOption update={onPress} options={options} activeOption={1} title="title" />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render options', () => {
    const { queryByText } = render(<CalculatorMultiOption update={onPress} options={options} activeOption={1} title="title" />)
    expect(queryByText('title')).not.toBeNull()
    expect(queryByText('TITLE_1')).not.toBeNull()
    expect(queryByText('TITLE_2')).not.toBeNull()
    expect(queryByText('subtitle_1')).not.toBeNull()
    expect(queryByText('subtitle_2')).not.toBeNull()
  })

  it('should handle update event', async () => {
    const { getByText } = render(<CalculatorMultiOption update={onPress} options={options} activeOption={1} title="title" />)
    expect(onPress).not.toHaveBeenCalled()
    await fireEvent.press(getByText('TITLE_2'))
    expect(onPress).toHaveBeenCalledWith(2)
  })
})
