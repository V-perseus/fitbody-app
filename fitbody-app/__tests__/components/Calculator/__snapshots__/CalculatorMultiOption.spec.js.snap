// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculatorMultiOption renders snapshot correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "marginBottom": 20,
      "marginTop": 20,
      "paddingLeft": 20,
      "paddingRight": 20,
      "width": 750,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#333333",
        "fontFamily": "OpenSans-Semibold",
        "fontSize": 16,
      }
    }
  >
    title
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 2,
        "flexDirection": "row",
        "flexGrow": 0,
        "flexWrap": "wrap",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#fa8dd5",
          "borderRadius": 30,
          "height": 50,
          "marginBottom": 10,
          "marginTop": 10,
          "opacity": 1,
          "width": 322.5,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "height": 50,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#333333",
                "fontFamily": "OpenSans-Semibold",
              },
              Object {
                "color": "#ffffff",
                "fontSize": 12,
              },
            ]
          }
        >
          TITLE_1
        </Text>
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 10,
            }
          }
        >
          subtitle_1
        </Text>
      </View>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "borderColor": "#dedede",
          "borderRadius": 30,
          "borderWidth": 1,
          "height": 50,
          "marginBottom": 10,
          "marginTop": 10,
          "opacity": 1,
          "width": 322.5,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "column",
            "height": 50,
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#333333",
                "fontFamily": "OpenSans-Semibold",
              },
              Object {
                "color": "#333333",
                "fontSize": 12,
              },
            ]
          }
        >
          TITLE_2
        </Text>
        <Text
          style={
            Object {
              "color": "#bbbbbb",
              "fontSize": 10,
            }
          }
        >
          subtitle_2
        </Text>
      </View>
    </View>
  </View>
</View>
`;
